bundle/edu.gemini.programexport.servlet/src/main/scala/edu/gemini/programexport/servlet/ProgramExportServlet.scala

```diff
@@ -0,0 +1,209 @@
+package edu.gemini.programexport.servlet
+
+import java.io.BufferedOutputStream
+import java.security.Principal
+import java.util.logging.{Level, Logger}
+import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
+import edu.gemini.pot.spdb.IDBDatabaseService
+import edu.gemini.spModel.core.{ProgramId, SPProgramID}
+import scalaz._
+import Scalaz._
+import edu.gemini.pot.sp.{ISPGroup, ISPNode, ISPObsComponent, ISPObservation, ISPProgram, SPComponentType}
+import edu.gemini.shared.util.immutable.ScalaConverters._
+import edu.gemini.spModel.data.YesNoType
+import edu.gemini.spModel.gemini.obscomp.{SPProgram, SPSiteQuality}
+import edu.gemini.spModel.obs.SPObservation
+import edu.gemini.spModel.obscomp.{ProgramNote, SPGroup, SPNote, SchedNote}
+import edu.gemini.spModel.target.env.TargetEnvironment
+import edu.gemini.spModel.util.SPTreeUtil
+
+import javax.servlet.ServletException
+import scala.collection.JavaConverters._
+
+final case class ProgramExportServlet(odb: IDBDatabaseService, user: Set[Principal]) extends HttpServlet {
+
+  import ProgramExportServlet._
+
+  override def doPost(request: HttpServletRequest, response: HttpServletResponse): Unit =
+    doRequest(request, response)
+
+  override def doGet(request: HttpServletRequest, response: HttpServletResponse): Unit =
+    doRequest(request, response)
+
+  def doRequest(request: HttpServletRequest, response: HttpServletResponse): Unit =
+  // Attempt to retrieve the program name: ProgramID instead?
+    ParamParser(request).id.fold(t => throw new ServletException(t), build(_, response))
+
+  // Build the JSON in a bottom-up approach starting at ISPProgram.
+  // Traverse a node's children and on the way back up, add the information for the node.
+  // Right now we just return an empty String: whether to use Strings or Argonaut to build the
+  // JSON is a decision I'm leaving to the end.
+  def build(id: SPProgramID, response: HttpServletResponse): Unit =
+    Option(odb.lookupProgramByID(id)) match {
+      case Some(ispProgram) =>
+        val out: BufferedOutputStream = new BufferedOutputStream(response.getOutputStream)
+        response.setContentType("text/plain")
```

Comment:
Should the response type be  like `application/json`?

Reply:
I know how JSON is structured but this is my first time actually using it; I didn't realize there was a type `application/json`.

Reply:
Thanks for that tip. Fixed.